# PROFUNDIZANDO LOS TIPOS DE DATOS EN PYTHON

En Python existen diferentes tipos de datos básicos, los cuales está definidos ya por defecto: enteros (int), flotantes (float), complejos (complex), cadenas de caracteres (str), booleanos (bool) que es un subconjunto de los enteros, listas (list), tuplas (tuple), diccionarios (dict), rangos (range) y conjuntos (set).

Además de estos tipos de datos básicos, se pueden importar bibliotecas que definen nuevas "clases" las cuales introducen nuevos tipos de datos, por ejemplo la biblioteca datetime introduce los tipos de datos: date, time, datetime, timedelta, tzinfo y timezone o la biblioteca math que tiene a sin, abs, factorial, log, etc.

Por último, se puede escribir un código en el cual se definan nuevas "clases" para introducir nuevos tipos de datos que necesite para el desarrollo de su software. Más adelante, en el apartado de "clases y objetos" se verá a detalle cómo definir nuevos tipos de datos a los que se llamará objetos.


## TIPOS DE DATOS NUMÉRICOS

### ENTEROS (int)

Corresponden a los números enteros positivos y negativos, e incluyen al 0. 

In [ ]:
# int?
help(int)

In [ ]:
a = 10
print(a,'\n')
print(id(a),'\n')
print(type(a))
# la función type devuelve la clase o tipo del objeto.
# Observe que en la útima línea se ha utilizado una función dentro de otra función.

Si multiplica dos datos enteros, el resultado será otro dato entero.

In [ ]:
x = 5
X = 10
r = x*X
print(r)
print(type(x))

### VEAMOS ALGUNOS MÉTODOS DEL INT
**int.bit_length( ):**
Retorna el número de bits necesarios para representar un número entero, excluyendo el bit de signo y los ceros a la izquierda

**int.bit_count( ):**
Devuelve el número de unos en la representación binaria del valor absoluto del entero. Esto también se conoce como recuento de población.

**int.as_integer_ratio( ):**
Retorna una pareja de números enteros cuya proporción es igual a la del numero entero original, y con un denominador positivo. En el caso de números enteros, la proporción siempre es el número original y 1 en el denominador.


In [1]:
# APLICANDO ALGUNOS MÉTODOS DEL OBJECTO INT
n=12
x=n.bit_length()
#y=n.bit_count() válido en la version 3.10
z=n.as_integer_ratio()
print(bin(n),'\n')
print(x,'\n')
#print(y,'\n')
print(z,'\n')


0b1100 

4 

(12, 1) 



### FLOTANTES (float)

Corresponden a los números que forman parte del conjunto de los reales. Se expresan como decimales o fracciones. 

In [ ]:
help(float)

In [ ]:
y = 0.52
z = 8/3
a = 9.7589357532
print(type(y))
print(type(z))
print(type(a))

Python tiene soporte para manejar diferentes tipos de datos numéricos, puede operar con enteros y flotantes.

In [ ]:
r2 = y + z - a
print(r2)
r3 = y*x
print(r3)

In [ ]:
r = 5.5 * 4
print(type(r))

### METODOS DEL FLOAT
**float.is_integer( ):**
Retorna True si el valor en coma flotante se puede representar sin perdida con un número entero, y False si no se puede.

**float.hex( ):**
Retorna la representación de un valor en coma flotante en forma de cadena de texto en hexadecimal. Para números finitos, la representación siempre empieza con el prefijo 0x, y con una p justo antes del exponente.

**float.as_integer_ratio( ):**
Retorna una pareja de números enteros cuya proporción es exactamente igual que la del valor en punto flotante original, con un denominador positivo.

In [2]:
b=34.896
c=3.9999999
d=3.9999999999999999999999999999999999999
x=b.is_integer()
y=b.as_integer_ratio()
z=b.hex()
m=c.is_integer()
n=d.is_integer()
print(x,'\n')
print(y,'\n')
print(z,'\n')
print(type(z),'\n')
print(m,'\n')
print(n,'\n')

False 

(2455587696823763, 70368744177664) 

0x1.172b020c49ba6p+5 

<class 'str'> 

False 

True 



El resultado de operar datos flotantes con datos enteros resultará un dato flotante

In [3]:
r4 = 6*3.8 # 22.8
r5 = 5*4.4 # 22.0
print(r4,'\n')
print(r5,'\n')
print(type(r4),'\n')
print(type(r5))

22.799999999999997 

22.0 

<class 'float'> 

<class 'float'>


En la celda anterior se observa que el resultado 22.80 se muestra como 22.799999999999, por lo que se estará preguntando: ¿Porque no se muestra el valor como tal? o si se ¿Puedo limitar la cantidad de decimales?

Por defecto los números de tipo " float " están almacenados en un formato llamado IEEE-754, lo cual ocasiona que se expresen con una cantidad considerable de números decimales. Y es que el principal motivo es que los números flotante no se pueden representar en binario de forma exacta. Sin embargo, al mostrar sus datos como cadenas de caracteres puede formatear los datos con el método " .format( ) ", el cual tiene la siguiente sintaxis:

str.format(variable)

## MÉTODO DE FORMATEO ( str.format( ) )

*link:* https://docs.python.org/es/3/library/string.html#string-formatting

In [ ]:
format?

In [10]:
r4 = 6*3.8 # 22.8
print("Mi resultado de",r4,"es %.2f" %r4 )

Mi resultado de 22.799999999999997 es 22.80


In [6]:
#r4 = 6*3.8 -> 22.8
resultado = "Mi resultado es {:.2f}".format(r4)
print(resultado)

Mi resultado es 22.80


Dentro de la cadena de caracteres (str) se introducirán un par de llaves que indicarán el lugar en donde se introducirá la variable a al que se va a dar formato.

Dentro de los paréntesis de format se escribe la variable que reemplazará a las llaves de la cadena de caracteres.

Dentro de las llaves puede esciribir " : . Nf " donde N es el número de decimales que se mostrará en pantalla.

In [5]:
resultado1 = "Mi resultado es {}".format(r4)
resultado2 = "Mi resultado es {:.7f}".format(r4)
resultado3 = "Mi resultado es {:.5f}".format(100)

print(resultado1,'\n')
print(resultado2,'\n')
print(resultado3)

Mi resultado es 22.799999999999997 

Mi resultado es 22.8000000 

Mi resultado es 100.00000


Puede revisar la documentación oficial de Python sobre las limitaciones de los números "float" en el siguiente link https://docs.python.org/es/3/tutorial/floatingpoint.html

### COMPLEJOS (complex)

Corresponden a los números complejos, definidos por una parte real y una parte imaginaria. Con el fin de representar la parte imaginaria, la variable j está definida por defecto como el número imaginario "i" que representa a la raiz de -1.

In [ ]:
help(complex)

In [ ]:
complejo = 8 + 5j
print(id(complejo),'\n')
print(complejo,'\n')
print(type(complejo),'\n')

# METODO PARA COMPLEX
a=complejo.conjugate()
print(id(a),'\n')
print(a)

Recuerde que si quiere utilizar número complejos se recomienda no utilizar la variable j para almacenar otro datos.

## BOLEANOS (bool)

Los booleanos son: True (verdadero) y False (falso). Estos tipos de datos son muy útiles porque permiten realizar operaciones lógicas. 

In [ ]:
#help(bool)
bool?

In [ ]:
T = True
F = False
print(T)
print(type(T))
print(F)
print(type(F))

El resultado de comparar dos números es un booleano:

In [ ]:
b1 = 5 < 8
print(b1)

El resultado de operar de manera lógica dos booleanos es otro booleano:

In [ ]:
b2 = True or False
print(b2)

True es equivalente a 1 y False es equivalente a 0. Puede utilizar booleanos para realizar operaciones matemáticas:

In [17]:
r1 = (58 + 5) * True
print(r1)
r2 = (58 + 5) * False
print(r2)

63
0


### COMO EXPRESARIAMOS LAS SIGUIENTES OPERACIONES

In [18]:
# El número es múltiplo de 4 y tbm es negativo
    #a%4==0 and a<0

False

In [22]:
# Decidiste comprar un auto usado, pero debe cumplir ciertas condiciones:
# El kilometraje debe ser menor a 30000 y el modelo debe estar entre 2015 y 2017
    # m -> kilometraje
    # n -> modelo
    #m<30000 and ( n>2015 and n<2017 )

In [23]:
# Una agrupación tiene ciertos requisitos para que cualquiera ingrese
# debe tener más del 30% de estudios completos y no debe ser miembro de otro grupo
    # porcentaje_ec><30 and not(miembro_agrupacion) 

## TIPOS DE DATOS ESTRUCTURA O DE COLECCIÓN

[enlace del modelo de los datos](https://docs.python.org/es/3/reference/datamodel.html?highlight=mutabilidad)

**LOS TIPOS INTEGRADOS BÁSICOS EN PYTHON**

| TIPO | OBJETO |
| :-------- | :------------- |
| Secuencias Mutable | list |
| Secuencias Inmutable | str, tuple |
| Conjuntos | set |
| Mapeos | dict |
| Invocable | funciones y/o métodos|
| Módulo | archivos .py|

### CADENA DE CARACTERES (str)

Se definen encerrando un texto con comillas simples, dobles o triples.

Las comillas triples se utilizan para mensajes largos que tienen una exentesión de dos o más líneas.

In [ ]:
#help(str)
str?

In [34]:
mensaje = "hola"
letra = "a"
correo = """
            Estimado, este es un ejemplo
            Buenas noches .......
            ....................
            ....................
            ...................
            Saludos
        """
print(mensaje)
print(type(mensaje))
print(letra)
print(type(letra))
print(correo)
print(type(correo))

hola
<class 'str'>
a
<class 'str'>

            Estimado, este es un ejemplo
            Buenas noches .......
            ....................
            ....................
            ...................
            Saludos
        
<class 'str'>


Observe que un único caracter también es de la clase str

Las listas es un tipo de dato secuencia y por ende soporta la indexación y por ende el slacing. También por eso se puede estudiar su mutabilidad.

In [37]:
g = 'La imaginación es más importante que el conocimiento'
print(g[::2],'\n')
print(g[3:14] +' y '+ g[40:],'\n')

L mgncó smsiprat u lcncmet 

imaginación y conocimiento 



In [38]:
# POR QUE DECIMOS QUE LAS LISTAS ES UN DATO DE TIPO SECUENCIAL PERO INMUTABLE
mensaje = "hola"
mensaje[0] = 'H'
print(mensaje)

TypeError: 'str' object does not support item assignment

### LISTAS (list)

Son un conjunto de datos ordenados. Se definen encerrando los datos entre corchetes "[ ]" y separándolos por comas.  

In [ ]:
#help(list)
list?

In [ ]:
t=[]
print(t,'\n')
print(id(t),'\n')
print(type(t),'\n')

In [ ]:
lista1 = [False, 5, 5]
print(lista1,'\n')
print(type(lista1))

Las listas pueden contener cualquier tipo de dato dentro de sí. Incluso se pueden crear listas dentro de listas. 

In [ ]:
lista2 = [8, 7, 9.875, 20, "hola", True, [5,8,10], lista1]
print(lista2,'\n')
print(type(lista2))

In [ ]:
w = [2, 4, True, 6]
print(w,'\n')
print(id(w),'\n')
w = [2, 4, True, 6, (5,7,False)]
print(w,'\n')
print(id(w),'\n')

### INDEX Y SLACING

Los datos de las listas están enumerados, de tal forma que podemos acceder a ellos a través de la notación de los corchetes. Esta notación consiste en escribir el nombre de la lista y luego entre corchetes escribir el número de orden del dato.

In [24]:
# TENEMOS INDEXACIÓN POSITIVA Y NEGATIVA
q = [False, 9+4j , 4.6, 'Viva el Perú']
print(q,'\n')
print(id(q),'\n')
Q = q[1:]
print(Q,'\n')
print(id(Q),'\n')

[False, (9+4j), 4.6, 'Viva el Perú'] 

2121155896000 

[(9+4j), 4.6, 'Viva el Perú'] 

2121155908800 



In [ ]:
# lista1 = [False, 5, 5]
# lista2 = [8, 7, 9.875, 20, "hola", True, [5,8,10], lista1]
dato1 = lista2[5]
dato2 = lista2[2:5]
dato3 = lista2[:7]
dato4 = lista2[3:]
dato5 = lista2[6][1]
dato6 = lista2[4][2]
dato7 = lista2[::2]
dato8 = lista2[::3]
dato9 = lista2[-1][-3]
print(dato1,'\n')
print(dato2,'\n')
print(dato3,'\n')
print(dato4,'\n')
print(dato5,'\n')
print(dato6,'\n')
print(dato7,'\n')
print(dato8,'\n')
print(dato9,'\n')
print('------------------------------------ TYPES ------------------------------------------')
print(type(dato1),'\n')
print(type(dato2),'\n')
print(type(dato3),'\n')
print(type(dato4),'\n')
print(type(dato5),'\n')
print(type(dato6),'\n')
print(type(dato7),'\n')
print(type(dato8),'\n')
print(type(dato9),'\n')

In [25]:
lista3 = [5.7, 7.5, "lennon", False]
print(id(lista3),'\n')
print(lista3,'\n')
print(type(lista3),'\n')

s=lista3[2][2:4]
print(s,'\n')
print(id(s),'\n')
print(type(s))

2121155909056 

[5.7, 7.5, 'lennon', False] 

<class 'list'> 

nn 

2121155857072 

<class 'str'>


In [26]:
# POR QUE DECIMOS QUE LAS LISTAS ES UN DATO DE TIPO SECUENCIAL PERO MUTABLE
lista3 = [5.7, 7.5, "lennon", False]
lista3[0]='Hendrix'
print(lista3)

['Hendrix', 7.5, 'lennon', False]


### TUPLAS (tuple)

Son un conjunto de datos numerados, muy similares a las listas. Se diferencia de las listas porque las tuplas son estáticas e inmutables, mientras que las listas son dinámicas y mutables.

In [ ]:
#help(tuple)
tuple?

In [ ]:
# TUPLA DE UN SOLO VALOR
v=()
print(v,'\n')
print(type(v),'\n')
print(id(v),'\n')

e=(5,)
print(e,'\n')
print(type(e),'\n')
print(id(e))

In [ ]:
tupla1 = (4, 8, 9, 10)
print(id(tupla1),'\n')
print(tupla1,'\n')
print(type(tupla1))

### INDEX Y SLACING

In [ ]:
tupla2 = (5, 7.5, "hola", False, tupla1)
print(id(tupla2),'\n')
print(tupla2,'\n')
print(type(tupla2),'\n')

s=tupla2[:4]
print(s,'\n')
print(id(s),'\n')
print(type(s))

In [27]:
tupla3 = (4, 8, 9, 10,[3,'HOLA',True])
print(tupla3[0],'\n')
print(tupla3[::2],'\n')
print(tupla3[4][1][:2],'\n')
print(tupla3[-1][-2])

4 

(4, 9, [3, 'HOLA', True]) 

HO 

HOLA


In [28]:
# POR QUE DECIMOS QUE LAS LISTAS ES UN DATO DE TIPO SECUENCIAL PERO INMUTABLE
tupla3 = (5.7, 7.5, "lennon", False)
tupla3[0]='Hendrix'
print(tupla3)

TypeError: 'tuple' object does not support item assignment

### DICCIONARIOS (dict)

Son un conjunto ordenando de pares de claves y valores. Las claves, llamadas "keys", son los nombres que se utilizan para etiquetar a los valores, también llamados "values".

Se definen escribiendo parejas de claves y valores encerradas entre llaves "{ }" y separadas por comas. Las claves y valores se separan por dos puntos " : ". Observe el ejemplo:

In [ ]:
#help(dict)
dict?

In [ ]:
diccionario = {"nombre": "Francisco", "apellido": "Arevalo", "nota": 20}
print(diccionario,'\n')
print(type(diccionario))

Las claves pueden ser cadenas de caracteres o datos numéricos , mientras que los valores pueden ser cualquier tipo de dato.

In [29]:
diccionario = {"nombre": "Francisco", "apellido": "Arevalo", "nota": 20}
dic = {1: "Francisco", 2: "Arevalo", 3: 20}
print(diccionario,'\n')
print(dic,'\n')
print(type(diccionario),'\n')
print(type(dic))

{'nombre': 'Francisco', 'apellido': 'Arevalo', 'nota': 20} 

{1: 'Francisco', 2: 'Arevalo', 3: 20} 

<class 'dict'> 

<class 'dict'>


Si un diccionario es extenso puede ordenarlo de mejor manera si lo separa por comas y saltos de linea.

In [ ]:
diccionario2 = {
    "ejemplo": "este es un ejemplo",
    "string": "hola",
    "int": 58,
    "float": 1.5856,
    "bool": True,
    "lista": [10,20,30,"cuarenta", True],
    "tupla": (4,5,10.525,"cincuenta", False),
    "diccionario": {"lista2":[1,2,3], "tupla2": (4,5)},
}
print(diccionario2,'\n')
print(id(diccionario2),'\n')
print(type(diccionario2),'\n')
print('---------------------- CREANDO OTRO OBJETO -----------------------------\n')
s = diccionario2["diccionario"]
print(id(s),'\n')
print(s["lista2"][::2])

Puede solicitar un dato del diccionario con la notación de los corchetes: 

nombreDelDiccionario[clave] 

Lo anterior devolverá como respuesta el valor de dicha clave. Los diccionario no soportan la indexación y/o slacing, justamente por que los diccionarios no son tipos de datos secuencia.

In [ ]:
print(diccionario2["bool"])
print(diccionario2["diccionario"]['tupla2'])

La notación de corchetes es muy útil para solicitar datos que se encuentran ordenados bajo estructuras complejas.

In [ ]:
print(diccionario2["diccionario"]["lista2"][:2])

In [30]:
# AHORA A LOS DATOS TIPO SECUENCIA SE LES PUEDE CLASIFICAR SEGÚN LA MUTABILIDAD
diccionario = {"nombre": "Francisco", "apellido": "Arevalo", "nota": 20}
diccionario['nombre']='jose'
print(diccionario,'\n')


{'nombre': 'jose', 'apellido': 'Arevalo', 'nota': 20} 



### CONJUNTOS (set)

In [ ]:
conjunto_vacio = set() # no puede ser {}, ya que este es un dict vacio
print(conjunto_vacio,'\n')
print(type(conjunto_vacio))

In [31]:
# EJEMPLO1
a = {2, 4, 7, 8, 8, 9, 17.5}
print(a,'\n')
print(id(a),'\n')
print(type(a),'\n')

print('\n','---------------------------------------------------------------','\n')

# EJEMPLO 2
b = {4,True,'lep zeppelin',6,6, (False,0),4,(False,0)}
print(b,'\n')
print(id(b),'\n')
print(type(b),'\n')

{17.5, 2, 4, 7, 8, 9} 

2121156722976 

<class 'set'> 


 --------------------------------------------------------------- 

{True, 4, 6, (False, 0), 'lep zeppelin'} 

2121156723648 

<class 'set'> 



Los conjuntos no tiene propiedades como la indexación o slacing. Ahora los conjuntos no es un tipo de dato secuencia pero es un caso particular porque según la documentación esta es inmutable y hay que tomarla como tal.

## MÉTODOS EN LA SALIDAD CON PRINT

In [12]:
# PRIMER MÉTODO
nombre = 'Miguel Stephano Albornoz Rivera'
edad = 21
pais = 'Perú'
print('Hola, me presento soy:',nombre,', tengo',edad,'años y soy del',pais)

Hola, me presento soy: Miguel Stephano Albornoz Rivera , tengo 21 años y soy del Perú


In [14]:
# SEGUNDO MÉTODO
nombre = 'Miguel Stephano Albornoz Rivera'
edad = 25
pais = 'Chile'
print(f'Hola, me presento soy: {nombre}, tengo {edad} años y soy de {pais}.')

Hola, me presento soy: Miguel Stephano Albornoz Rivera, tengo 25 años y soy de Chile.


In [16]:
# TERCER MÉTODO
nombre = 'Miguel Stephano Albornoz Rivera'
edad = 32
pais = 'España'
print('Hola, me presento soy: {}, tengo {} años y soy de {}.'.format(nombre,edad,pais))

Hola, me presento soy: Miguel Stephano Albornoz Rivera, tengo 32 años y soy de España.


## IMPORTANDO MÓDULOS

In [ ]:
# import math
import math as mt
# from math import sqrt
# from math import log
# from math import fabs
# from math import factorial

In [ ]:
# OPERACIONES COMBINADAS
X = mt.log(100) + pow(mt.sqrt(4),8) - mt.factorial(5) * mt.fabs(-12)
print(x)

In [ ]:
# CALCULAR EL AREA DE UN CIRCULO
# A = pi * r^2
r=3.5
A = mt.pi * pow(r,2)
print('el area del cieculo es: {:.3f}'.format(A))

In [ ]:
# CALCULAR FUNCIONES TRIGONOMÉTRICAS
x = 30 # tenemos que convertir a radianes
z = mt.sin((x * mt.pi) / 180)
print(' el seno de',x,'es:',z)

In [ ]:
help(mt)


## INPUT( )
La función input( ) regresa o retorna siempre una cadena, es semejante a un dispositivo de entrada como el teclado.

In [ ]:
input?
#help(input)

In [ ]:
s = input()
print(type(s),'\n')
print(id(s))

## PRIMEROS SCRIPTS

In [ ]:
a = input('ingrese la cantidad de bicicletas que tiene: ')
print(type(a))
b = eval(a)
print(b)
print(type(b))

In [1]:
# saca el valor absoluto de x
e = abs(-5)
e

5

In [2]:
# eleva la potencia 2**15
w = pow(2,15)
w

32768

In [4]:
# eleva la potencia 2**6 y luego saca el resto de 3
s = pow(2,6,3)
s

1

In [8]:
x = divmod(6,3)
print(x)
y = divmod(17,3)
print(y)

(2, 0)
(5, 2)


In [24]:
#s = round(14.5,2)
s= round(14.537849930,0)
s

15.0

In [25]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


enlace: [PEP8](https://www.python.org/dev/peps/pep-0008/)

## PARA PRACTICAR:

a) Solicite el valor 58 de la variable diccionario2

In [2]:
diccionario2 = {
    "ejemplo": "este es un ejemplo",
    "string": "hola",
    "int": 58,
    "float": 1.5856,
    "bool": True,
    "lista": [10,20,30,"cuarenta", True],
    "tupla": (4,5,10.525,"cincuenta", False),
    "diccionario": {"lista2":[1,2,3], "tupla2": (4,5)},
}

In [3]:
diccionario2['int']

58

b) Solicite el valor 5 de clave "tupla2" que se encuentra en la clave "diccionario" de la variable diccionario2

In [5]:
diccionario2['tupla'][1]

5

c) Cree un variable con cada tipo de dato presentado en este documento (son 8 en total) y utilice la función type para mostrar en pantalla el tipo de dato de cada variable. 

In [16]:
a=5
print(a,' -> ',type(a),'\n')
b=8.0
print(b,' -> ',type(b),'\n')
c=True
print(c,' -> ',type(c),'\n')
d=3 + 5j
print(d,' -> ',type(d),'\n')
e=4,
print(e,' -> ',type(e),'\n')
f=[2,False,(9,0)]
print(f,' -> ',type(f),'\n')
g={'lunes':1,'martes':2,'miercoles':3,'jueves':4,'viernes':5}
print(g,' -> ',type(g),'\n')
h={3,3,0,'hola'}
print(h,' -> ',type(h),'\n')

5  ->  <class 'int'> 

8.0  ->  <class 'float'> 

True  ->  <class 'bool'> 

(3+5j)  ->  <class 'complex'> 

(4,)  ->  <class 'tuple'> 

[2, False, (9, 0)]  ->  <class 'list'> 

{'lunes': 1, 'martes': 2, 'miercoles': 3, 'jueves': 4, 'viernes': 5}  ->  <class 'dict'> 

{0, 3, 'hola'}  ->  <class 'set'> 



d) Imprima en pantalla el resultado de la siguiente operación, expresado con 3 decimales de precisión: 

(38 x 3 - 45) x (1/2)

El símbolo "x" expresa multiplicación

In [7]:
print('{} {:.3f}'.format('el resultado de (38 * 3 - 45) * (1/2) es:',(38*3-45)*(1/2)))


el resultado de (38 * 3 - 45) * (1/2) es: 34.500
